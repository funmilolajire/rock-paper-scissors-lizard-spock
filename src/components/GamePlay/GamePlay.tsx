import { Icon } from "../Icon/Icon"
import { GameOver } from '../GameOver/GameOver';
import { useHouseFighter, usePlayerFighter, fighters } from "../../state";
import { useEffect, useCallback } from 'react';
import { outcome } from "../../state/utils/rules";
import { useState } from 'react';

export const GamePlay = () => {
    const [gameOver, setGameOver] = useState('')
    //get player and house fighters
    const playerFighter = usePlayerFighter();
    const houseFighter = useHouseFighter();
    const player = playerFighter.get();
    const house = houseFighter.get();
    const randomFighter = useCallback(() => {
        const randomNumber = Math.floor(Math.random() * fighters.length)
        let fighter = fighters[randomNumber]
        return houseFighter.select(fighter);
    }, [])
    useEffect(() => {
        houseFighter.select('')
        let houseTimeout: any;
        if (player) {
            houseTimeout = setTimeout(() => {
                randomFighter()
                setGameOver(outcome())
            }, 2000)
        }
        return () => {
            clearTimeout(houseTimeout)
        }
    }, [player, randomFighter])
    useEffect(() => {
        setGameOver('')
        const playerClasses = document.querySelector('[data-player-fighter]')?.classList;
        const houseClasses = document.querySelector('[data-house-fighter]')?.classList;
        if (playerClasses?.contains('winner')) {
            playerClasses?.remove('winner')
        }
        if (houseClasses?.contains('winner')) {
            houseClasses?.remove('winner')
        }
    }, [player, randomFighter])

    return (
        <div className="GamePlay" data-testid="GamePlay">
            <section className="Player">
                <h3>You Picked</h3>
                <div data-player-fighter className="Fighter">
                    {player && <Icon iconName={player} />}
                    <span className="Empty"></span>
                </div>
            </section>
            {gameOver && <GameOver gameOver={gameOver} />}
            <section className="House">
                <h3>The house picked</h3>
                <div data-house-fighter className="Fighter">
                    {house && <Icon iconName={house} />}
                    <span className="Empty"></span>
                </div>
            </section>
        </div>
    )
}