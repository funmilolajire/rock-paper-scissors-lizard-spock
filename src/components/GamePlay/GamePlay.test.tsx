import { mount } from 'enzyme';
import { GamePlay } from './GamePlay';

describe('gameplay component', () => {
    let component = mount(<GamePlay />)
    test('make a snapshot of gameplay component', () => {
        expect(component).toMatchSnapshot()
    })
    test('winner class is absent when component loads', () => {
        const playerRing = component.find('[data-player-fighter]').hasClass('winner')
        const houseRing = component.find('[data-house-fighter]').hasClass('winner')
        expect(playerRing).toBeFalse()
        expect(houseRing).toBeFalse()
    })
})
