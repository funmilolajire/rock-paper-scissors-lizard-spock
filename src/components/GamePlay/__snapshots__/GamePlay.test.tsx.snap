// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gameplay component make a snapshot of gameplay component 1`] = `
<GamePlay>
  <div
    className="GamePlay"
    data-testid="GamePlay"
  >
    <section
      className="Player"
    >
      <h3>
        You Picked
      </h3>
      <div
        className="Fighter"
        data-player-fighter={true}
      >
        <span
          className="Empty"
        />
      </div>
    </section>
    <section
      className="House"
    >
      <h3>
        The house picked
      </h3>
      <div
        className="Fighter"
        data-house-fighter={true}
      >
        <span
          className="Empty"
        />
      </div>
    </section>
  </div>
</GamePlay>
`;
